src/analyses/base.ml

```diff
@@ -2093,16 +2094,30 @@ struct
             set ~ctx (Analyses.ask_of_ctx ctx) gs st s1_a s1_typ (VD.top_value (unrollType s1_typ))
         end
       (* else compute value in array domain *)
-      else 
+      else
         let lv_a, lv_typ = match lv with
           | Some lv_val -> eval_lv (Analyses.ask_of_ctx ctx) gs st lv_val, Cilfacade.typeOfLval lv_val
           | None -> s1_a, s1_typ in
-        begin match get (Analyses.ask_of_ctx ctx) gs st s1_a None, get (Analyses.ask_of_ctx ctx) gs st s2_a None with
+        begin match (get (Analyses.ask_of_ctx ctx) gs st s1_a None), get (Analyses.ask_of_ctx ctx) gs st s2_a None with
           | Array array_s1, Array array_s2 -> set ~ctx (Analyses.ask_of_ctx ctx) gs st lv_a lv_typ (op_array array_s1 array_s2)
-          | Array array_s1, _ when s2_typ = charPtrType -> 
+          | Array array_s1, _ when s2_typ = charPtrType ->
             let s2_null_bytes = List.map CArrays.to_null_byte_domain (AD.to_string s2_a) in
             let array_s2 = List.fold_left CArrays.join (CArrays.bot ()) s2_null_bytes in
             set ~ctx (Analyses.ask_of_ctx ctx) gs st lv_a lv_typ (op_array array_s1 array_s2)
+          | Bot, Array array_s2 ->
+            (* If we have bot inside here, we assume the blob is used as a char array and create one inside *)
+            let size = ctx.ask (Q.BlobSize s1) in
+            let s_id = ValueDomainQueries.ID.unlift (Fun.id) size in
+            let empty_array = CArrays.make (ID.cast_to (Cilfacade.ptrdiff_ikind ()) s_id) (Int (ID.top_of IChar)) in
+            set ~ctx (Analyses.ask_of_ctx ctx) gs st lv_a lv_typ (op_array empty_array array_s2)
+          | Bot , _ when s2_typ = charPtrType ->
+            (* If we have bot inside here, we assume the blob is used as a char array and create one inside *)
+            let size = ctx.ask (Q.BlobSize s1) in
+            let s_id = ValueDomainQueries.ID.unlift (Fun.id) size in
+            let empty_array = CArrays.make (ID.cast_to (Cilfacade.ptrdiff_ikind ()) s_id) (Int (ID.top_of IChar)) in
+            let s2_null_bytes = List.map CArrays.to_null_byte_domain (AD.to_string s2_a) in
+            let array_s2 = List.fold_left CArrays.join (CArrays.bot ()) s2_null_bytes in
+            set ~ctx (Analyses.ask_of_ctx ctx) gs st lv_a lv_typ (op_array empty_array array_s2)
```

Comment:
When running Goblint now on some programs of the coreutils, it often crashes because of the exception `Queries.ID.unlift`, I suppose that is related to `s_id`.

Reply:
Yes, maybe we can handle this by catching and creating an `ID.top ()`? But still weird that `ValueDomainQueries.ID` would not be an `ID.t` here...

Reply:
Might make sense to log for which element we call Blobsize here?

Reply:
Or do you think this could be the same issue you had with blobs @mrstanb ?

Reply:
> Or do you think this could be the same issue you had with blobs @mrstanb ?

I'm not entirely sure what's going on here, but I think it could be a similar issue. In my case `BlobSize` was always giving `top` before we introduced the changes for `BlobSize`. Might be worth checking if this PR's branch is up-to-date with the `BlobSize` changes that should already be on `master`?

In this case `unlift` is probably complaining due to `bot` or `top`. If `BlobSize` is giving `bot` as a result, it might be worth checking if there's some offset in the abstract address set. In that case `BlobSize` will answer with `bot`.

Reply:
@nathanschmidt Were you able to take a look? If yes, I could do a final review tomorrow and then hopefully merge this.

Reply:
@michael-schwarz The branch is up to date with master. As you suggested, I caught the exception and created a top element. I suppose it's not ideal and should be fixed in the future, but it does the trick for now. I've also added some further regression tests after looking at coverage reviews, so feel free to review tomorrow!